### PR TITLE
Added peaks subtyping

### PR DESCRIPTION
**Pull request by Jianyu010**
_Before you submit this PR: make sure to put all operations-related information in a wiki-note, a PR should be about code and is publicly accessible_

## What does the code in this PR do / what does it improve?

enable processing for subtypings. See the note [here](https://xe1t-wiki.lngs.infn.it/doku.php?id=jlong:peak_subtyping_study)

## Can you briefly describe how it works?

Added the corresponding plugin in `peaks` directory.

## Can you give a minimal working example (or illustrate with a figure)?

_Please include the following if applicable:_
  - [ ] _Update the docstring(s)_
  - [ ] _Update the documentation_
  - [ ] _Tests to check the (new) code is working as desired._
  - [ ] _Does it solve one of the open issues on github?_

### _Notes on testing_
 - _Until the automated tests pass, please mark the PR as a draft._
 - _On the XENONnT fork we test with database access, on private forks there is no database access for security considerations._

All _italic_ comments can be removed from this template.

### CODE REVIEW

**Review by dachengx — CHANGES_REQUESTED** [review]

File: straxen/plugins/peaks/peaks_subtyping.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
+import numpy as np
+import strax
+import straxen
+import numba
+
+class PeaksSubtypes(strax.Plugin):
+    
+    '''
+    checkout https://xe1t-wiki.lngs.infn.it/doku.php?id=jlong:sr0_few_electron_ps2_picking
```
[the remark this answers is not in the document]
Comment (Jianyu010) [reply straxen/plugins/peaks/peaks_subtyping.py]: Done

File: straxen/plugins/peaks/peaks_subtyping.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    ref_se_gain = straxen.URLConfig(
+        default='bodega://se_gain?bodega_version=v1',
+        help='Nominal single electron (SE) gain in PE / electron extracted.')
+    
+    ref_se_span = straxen.URLConfig(
+        default='bodega://se_spread?bodega_version=v0',
+        help = "SE spread value"
+    )
+    
+    FV_r = straxen.URLConfig(
```
[the remark this answers is not in the document]
Comment (Jianyu010) [reply straxen/plugins/peaks/peaks_subtyping.py]: Deleted

File: straxen/plugins/peaks/peaks_subtyping.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        default = 50., type = float, #cm
+        help = "fiducial volume"
+    )
+    full_dt = straxen.URLConfig(
+        default = 2300e3, type = float,#ns
+        help = "maximum drift time from cathode"
+    )
+    max_area = straxen.URLConfig(
```
[the remark this answers is not in the document]
Comment (Jianyu010) [reply straxen/plugins/peaks/peaks_subtyping.py]: Deleted

File: straxen/plugins/peaks/peaks_subtyping.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        print('sorting S2s')
+        self.mark_s2s(peaks['time'],
+                      peaks['area'],
+                      self.mask,
+                      self.full_dt,
+                      self.ls2_threshold,
+                      self.other_ls2,
+                      self.s1_s2_window)
+        
+        '''self.register_livetime()
```
[the remark this answers is not in the document]
Comment (Jianyu010) [reply straxen/plugins/peaks/peaks_subtyping.py]: Deleted

File: straxen/plugins/peaks/peaks_subtyping.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+                        (s2s_a[whole_window_cond] < np.max([ls2_threshold,other_ls2 * max_val]))
+                        & (s2_mask[whole_window_cond] != 22)
+                        & (s2s_t[whole_window_cond] >= max_time)] = 272
+                    # assign the pS2 and S1
+                    s2_mask_tmp[s2s_t[whole_window_cond] == max_time] = 22
+                    s1_mask[i] = 11
+                    s2_mask[whole_window_cond] = s2_mask_tmp
+                else:
+                    #no pS2, loneS1 spotted
+                    whole_window_cond = (s2s_t > s1s_t[i]) & (s2s_t <= s1s_t[i] + full_dt)
```
Comment (dachengx) [inline straxen/plugins/peaks/peaks_subtyping.py:154]: Is there a potential bug? All other time window length is defined as `2 * full_dt`.

**Review by dachengx — CHANGES_REQUESTED** [review]

File: straxen/plugins/peaks/peaks_subtyping.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        If a small S2 is identified and a pS2 can be paired up with it, such small S2 is marked "fake S2". 
+        '''
+
+        undefined_mask = (mask == PeakSubtyping.Undefined)
+        dmask = mask[undefined_mask]
+        pt = peaks['time'][undefined_mask]  # time
+        pa = peaks['area'][undefined_mask]  # area
+
+        # load only the ones not being classified
+        while (dmask == PeakSubtyping.Undefined).sum() > 0:
```
[the remark this answers is not in the document]
Comment (Jianyu010) [reply straxen/plugins/peaks/peaks_subtyping.py:202]: 这个的目的是穷尽所有6 。目前我自己跑没遇到过问题。。这个是根据peak数量来的，上限有点难啊，9e9?

[the remark this answers is not in the document]
Comment (dachengx) [reply straxen/plugins/peaks/peaks_subtyping.py:202]: OK. I will add that.

File: straxen/plugins/peaks/peaks_subtyping.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+                    # do we have pS2 or a bunch of small S2s
+                    max_ind = np.nanargmax(pa[ps2_cand_cond])
+                    max_val = pa[ps2_cand_cond][max_ind]
+                    max_time = pt[ps2_cand_cond][max_ind]
+                    if max_val >= ls2_threshold:
+                        # a suspicious coupling spotted. label this s2 as fakeS2
+                        dmask_tmp = dmask[dmask == PeakSubtyping.Undefined]
+                        dmask_tmp[0] = PeakSubtyping.fakeS2
+                        dmask[dmask == PeakSubtyping.Undefined] = dmask_tmp
+                        whole_window_cond = (pt > start_s2_t) & (pt <= max_time + 2 * drift_time_max)
```
[the remark this answers is not in the document]
Comment (Jianyu010) [reply straxen/plugins/peaks/peaks_subtyping.py:232]: They should not be the same. the 1full_dt was for S1, and 2full_dt was for S2

**Review by dachengx — APPROVED** [review]

**Pull request merged**